### ISSUE TITLE
Can only build Python 3.8 virtual env

### ISSUE THREAD

**Issue by stub42** [canonical/charmcraft#205]
Charmcraft uses Python 3.8 to build the venv. This works fine for k8s charms (where the operator pod also runs Python 3.8), but if charms depend on Python libraries with C extensions then the build charm will only work under Focal.

For example, requiring `psycopg2-binary`, I end up with the following in my build charm:

```
 1262304  Defl:N   384090  70% 2020-12-01 18:41 23958baf  venv/psycopg2/_psycopg.cpython-38-x86_64-linux-gnu.so
```

The above will work fine if deployed with Focal, but will fail if the charm needs to support Xenial or Bionic. A possible work around is to install charmcraft using pip into a Bionic VM, so it runs with Python 3.6 and embeds Python 3.6 specific dependencies.

A more minor issue is the .pyc files are also embedded and Python 3.8 specific.

**Comment by stub42** [comment]
It might be best to avoid the problem entirely, and only support embedding pure Python dependencies. Charms requiring Python dependencies with C extensions could use native packages (apt), when dependencies on native packages are supported.

**Comment by facundobatista** [comment]
In the future the charms will be built on the specific required platforms, using Python and libraries versions from those platforms, so this will not be a problem.

**Comment by jameinel** [comment]
I believe the intent is to use the launchpad building infrastructure to
build platform dependent versions of the charm (certainly in the charmhub
interaction on the Juju side we have to specify both the os and series and
arch).

Certainly Charmhub is modelling each platform and architecture as a
separate charm binary that Juju needs to be aware of.

**Comment by sergiusens** [comment]
There's two scenarios to unfold here:

- building locally, where charmcraft can try and build for all the platforms it can from the defined target list
- integrating with launchpad, where launchpad from a given entry point (e.g.; architectures in snapcraft.yaml) dispatch builds (in destructive mode aka local host builds) to every known platform it has listed (with failure modes that we need to agree on for platforms it does not support)

**Comment by cjp256** [comment]
The snap is now required (and will be released in 1.1), which avoids the conflict of the charmcraft requirements vs project dependencies (the snap ships with its own interpreter which is not used for the charm's venv).
